crates/shrs_core/src/state.rs

```diff
@@ -78,31 +80,91 @@ impl<T: 'static> DerefMut for StateMut<'_, T> {
         self.value.downcast_mut().unwrap()
     }
 }
+
+// Potential errors that can occur when interacting with state store
+#[derive(Error, Debug)]
+pub enum StateError {
+    // TODO include the type in the error message
+    #[error("Value is missing")]
+    Missing,
+    #[error("Failed to borrow")]
+    Borrow,
+    #[error("Failed to borrow mut")]
+    BorrowMut,
+    #[error("Failed to downcast")]
+    Downcast,
+}
+
+// Global state store
 #[derive(Default)]
 pub struct States {
     states: HashMap<TypeId, RefCell<Box<dyn Any>>>,
 }
+
 impl States {
+
+    // Insert a new piece of state of given type into global state store, overriding previously
+    // existing values
+    // TODO should we allow overriding contents of state?
     pub fn insert<S: 'static>(&mut self, res: S) {
         self.states
             .insert(TypeId::of::<S>(), RefCell::new(Box::new(res)));
     }
-    pub fn remove<S>() {}
 
-    pub fn get_mut<S: 'static>(&self) -> RefMut<S> {
-        let s = self
-            .states
-            .get(&TypeId::of::<S>())
-            .expect("Value Missing")
-            .borrow_mut();
-        RefMut::map(s, |b| b.downcast_mut::<S>().unwrap())
+    // TODO this is potentially dangerous to allow arbitrary code to remove state
```

Comment:
@nithinmuthukumar thoughts on this too, should we even allow removal of state?

Reply:
Think its fine to allow removal of state but it has to come with optional di params, bevy allows Option<Res<T>> which is something we could also allow in the future

Reply:
We should also write in docs what types shouldnt be removed or add warnings if those types are removed

Reply:
optional state sounds cool, maybe lets open an issue to potentially implement that in the future? although i dont really see any uses cases of optional state currently